server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java

```diff
@@ -506,11 +523,76 @@ private List<UploadedIndexMetadata> writeIndexMetadataParallel(ClusterState clus
         return result;
     }
 
+    /**
+     * Invokes the index metadata upload listener.
+     */
+    private void invokeIndexMetadataUploadListeners(
+        List<IndexMetadata> newIndexMetadataList,
+        CountDownLatch latch,
+        List<Exception> exceptionList
+    ) {
+        for (IndexMetadataUploadListener listener : indexMetadataUploadListeners) {
+            // We are submitting the task for async execution to ensure that we are not blocking the cluster state upload
+            String listenerName = listener.getClass().getSimpleName();
+            String threadPoolName = listener.getThreadpoolName();
+            assert ThreadPool.THREAD_POOL_TYPES.containsKey(threadPoolName) && ThreadPool.Names.SAME.equals(threadPoolName) == false;
+            threadpool.executor(threadPoolName).execute(() -> {
+                try {
+                    listener.beforeNewIndexUpload(
```

Comment:
The method says `beforeNewIndexUpload` but we are uploading metadata in parallel, right?

Reply:
That's right. We are just listening to the part when the upload is happening. It does not mean that the upload has completed yet.

Reply:
@ashking94 it will be good to update the java docs as well that it will trigger ```beforeNewIndexUpload``` but it wouldnt wait for it finish.

Reply:
Let me rename this to `onNewIndexUpload`.